### PR TITLE
Fix the filtering of the complete state - Updating the unit tests

### PR DESCRIPTION
**Pull request by HorjuRares**
Issues: #

# Definition of Done

The PR shall be merged only if all items mentioned in [CONTRIBUTING.md](https://github.com/eclipse-ankaios/ankaios/blob/main/CONTRIBUTING.md#how-to-contribute) have been followed. In case an item is not applicable as described, please provide a short explanation in the description.

- [ ] All steps in [CONTRIBUTING.md](https://github.com/eclipse-ankaios/ankaios/blob/main/CONTRIBUTING.md#how-to-contribute) have been handled

### CODE REVIEW

File: agent/src/workload/workload_control_loop.rs [hunk 1]
```diff
@@ -1986,7 +1986,7 @@ mod tests {
 
         let workload_command_sender_clone = workload_command_sender.clone();
         tokio::spawn(async move {
-            tokio::time::sleep(tokio::time::Duration::from_millis(125)).await;
+            tokio::time::sleep(tokio::time::Duration::from_millis(250)).await;
```
Comment (krucod3) [inline agent/src/workload/workload_control_loop.rs:1989]: We should revert this as the failing tests were caused by the virtual machine.

File: agent/src/workload/workload_control_loop.rs [hunk 2]
```diff
@@ -2003,7 +2003,7 @@ mod tests {
         control_loop_state.state_checker = Some(old_state_checker);
 
         assert!(timeout(
-            Duration::from_millis(150),
+            Duration::from_millis(300),
```
Comment (krucod3) [inline agent/src/workload/workload_control_loop.rs:2006]: We should revert this as the failing tests were caused by the virtual machine.

File: server/src/ankaios_server.rs [hunk 3]
```diff
@@ -1047,23 +1048,25 @@ mod tests {
...
+
+        let current_complete_state = ank_base::CompleteState {
+            desired_state: Some(ank_base::State {
+                workloads: Some(workload_map),
                 ..Default::default()
-            },
+            }),
             ..Default::default()
         };
         let request_id = format!("{AGENT_A}@my_request_id");
```
Comment (krucod3) [inline server/src/ankaios_server.rs:1072]: This here is a candidate for a `generate_test_proto_complete_state`

File: common/src/commands.rs [hunk 4]
```diff
@@ -467,10 +480,16 @@ mod tests {
...
             .unwrap()
             .workloads
+            .as_mut()
+            .unwrap()
+            .workloads
             .insert(
                 WORKLOAD_NAME_1.into(),
                 ank_base::Workload {
-                    dependencies: vec![("dependency".into(), -1)].into_iter().collect(),
+                    // dependencies: vec![("dependency".into(), -1)].into_iter().collect(),
```
Comment (krucod3) [inline common/src/commands.rs:489]:
```suggestion
```

File: common/src/commands.rs [hunk 5]
```diff
@@ -263,29 +263,60 @@ mod tests {
...
-                desired_state: $expression::State {
+        (ankaios) => {
+            ankaios::CompleteState {
+                desired_state: ankaios::State {
                     api_version: "v0.1".into(),
-                    workloads: vec![("desired".into(), workload!($expression))]
-                        .into_iter()
-                        .collect(),
+                    // workloads: vec![("desired".into(), workload!(ankaios))].into_iter(),
+                    // .collect(),
```
Comment (krucod3) [inline common/src/commands.rs:271]:
```suggestion
```

File: ank/src/filtered_complete_state.rs [hunk 6]
```diff
@@ -79,6 +79,7 @@ pub struct FilteredWorkloadSpec {
 }
 
 impl From<ank_base::CompleteState> for FilteredCompleteState {
+    // TODO: create a conversion for common::objects::CompleteState
```
Comment (krucod3) [inline ank/src/filtered_complete_state.rs:82]:
```suggestion
```

File: ank/src/filtered_complete_state.rs [hunk 7]
```diff
@@ -136,3 +137,5 @@ impl From<ank_base::Workload> for FilteredWorkloadSpec {
         }
     }
 }
+
+// TODO: add helper function for generating filtered complete state
```
Comment (krucod3) [inline ank/src/filtered_complete_state.rs:141]: There are indeed multiple locations where such a function would be needed. Another possibility would be to add the `generate_test_proto_complete_state` fn and convert to filtered compete state in the tests (.into())

File: ank/src/cli_commands/server_connection.rs [hunk 8]
```diff
@@ -402,15 +405,20 @@ mod tests {
         );
         sim.will_send_response(
             REQUEST,
-            ResponseContent::CompleteState(Box::new(complete_state(WORKLOAD_NAME_1))),
+            ank_base::response::ResponseContent::CompleteState(
+                complete_state(WORKLOAD_NAME_1).into(),
```
Comment (krucod3) [inline ank/src/cli_commands/server_connection.rs:409]: Maybe we should change the complete_state() fn to return an ank_base complete state directly.

File: ank/src/cli_commands/server_connection.rs [hunk 9]
```diff
@@ -229,11 +229,10 @@ pub enum ServerConnectionError {
 mod tests {
     use std::collections::HashMap;
 
+    use super::ank_base::{self, UpdateStateSuccess};
+    // use api::ank_base::CompleteState;
```
Comment (krucod3) [inline ank/src/cli_commands/server_connection.rs:233]:
```suggestion
```

File: ank/src/cli_commands/get_workloads.rs [hunk 10]
```diff
@@ -139,7 +150,7 @@ mod tests {
         mock_server_connection
             .expect_get_complete_state()
             .with(eq(vec![]))
-            .return_once(|_| Ok(Box::new(test_data)));
+            .return_once(|_| Ok((ank_base::CompleteState::from(test_data)).into()));
```
Comment (krucod3) [inline ank/src/cli_commands/get_workloads.rs:153]: Maybe we can directly generate a proro (ank_base) complete state for test_data and use in instead of making 2 subsequent conversions.

**Review by krucod3 — COMMENTED** [review]
First part of review is done. I still need to review `runtime_manager.rs` and `get_state.rs`

I think we should add a `generate_test_proto_complete_state` function family that:
* generates an empty one with the proper api version
* accepts an array of tuples (impl Into<String>, ank_base::Workload)

File: ank/src/cli_commands/get_state.rs [hunk 11]
```diff
@@ -58,371 +55,53 @@ impl CliCommands {
...
-    };
+    use api::ank_base;
+    use common::{objects::generate_test_workload_spec_with_param, test_utils};
     use mockall::predicate::eq;
 
     use crate::{
-        cli::OutputFormat,
         cli_commands::{
-            get_state::{generate_compact_state_output, get_filtered_value, update_compact_state},
+            // get_state::{generate_compact_state_output, get_filtered_value, update_compact_state},
```
Comment (krucod3) [inline ank/src/cli_commands/get_state.rs:64]:
```suggestion
```

File: ank/src/cli_commands/server_connection.rs (hunk 8 above)
Comment (HorjuRares) [reply ank/src/cli_commands/server_connection.rs:409]: The instances where common::objects is created and needed greatly outnumber the ank_base ones. Instead I've proceeded to use proto_complete_state.

File: ank/src/cli_commands/get_workloads.rs [hunk 12]
```diff
@@ -135,11 +146,19 @@ mod tests {
             ),
         ]);
 
+        // let test_data = test_utils::generate_test_proto_complete_state(&[(
+        //     "name1",
+        //     test_utils::generate_test_proto_workload_with_param("agent_A", "runtime"),
+        // )]);
+        println!("{:?}", test_data);
+
```
Comment (krucod3) [inline ank/src/cli_commands/get_workloads.rs:154]: Please remove
```suggestion
```

File: ank/src/cli_commands/get_workloads.rs [hunk 13]
```diff
@@ -135,11 +146,19 @@ mod tests {
...
         let mut mock_server_connection = MockServerConnection::default();
         mock_server_connection
             .expect_get_complete_state()
             .with(eq(vec![]))
-            .return_once(|_| Ok(Box::new(test_data)));
+            .return_once(|_| Ok((ank_base::CompleteState::from(test_data)).into()));
+        // .return_once(|_| Ok(test_data.into()));
+
```
Comment (krucod3) [inline ank/src/cli_commands/get_workloads.rs:161]:
```suggestion
```

File: server/src/ankaios_server/server_state.rs [hunk 14]
```diff
@@ -424,32 +422,90 @@ mod tests {
...
+        //     runtime: Some(w1.runtime.clone()),
+        //     runtime_config: Some(w1.runtime_config.clone()),
+        //     control_interface_access: w1.control_interface_access.into(),
+        // },
+        //                 ),
+        //             ]),
+        //         }),
+        //     }),
+        //     workload_states: None,
+        // };
```
Comment (krucod3) [inline server/src/ankaios_server/server_state.rs:504]: Please remove the unneeded parts.
```suggestion
```

File: common/src/test_utils.rs [hunk 15]
```diff
@@ -33,6 +33,31 @@ pub fn generate_test_state_from_workloads(workloads: Vec<WorkloadSpec>) -> State
     }
 }
 
+// TODO: decide on conversion vs. dedicated generating fn
+// pub fn generate_test_proto_workloads_from_specs(
+//     workloads: Vec<WorkloadSpec>,
+// ) -> ank_base::WorkloadMap {
+//     todo!();
+// }
+
```
Comment (krucod3) [inline common/src/test_utils.rs:42]: Please remove:
```suggestion
```

File: api/src/lib.rs [hunk 16]
```diff
@@ -28,6 +28,11 @@ pub mod ank_base {
             }
         }
     }
+
+    // #[cfg(feature = "test_utils")]
+    // pub fn generate_test_complete_state(*necessary_params*) -> CompleteState {
+    //     todo!()
+    // }
```
Comment (krucod3) [inline api/src/lib.rs:35]: Please remove:
```suggestion
```

File: agent/src/workload/workload_control_loop.rs [hunk 17]
```diff
@@ -1986,7 +1986,7 @@ mod tests {
 
         let workload_command_sender_clone = workload_command_sender.clone();
         tokio::spawn(async move {
-            tokio::time::sleep(tokio::time::Duration::from_millis(125)).await;
+            tokio::time::sleep(tokio::time::Duration::from_millis(150)).await;
```
Comment (krucod3) [inline agent/src/workload/workload_control_loop.rs:1989]: Please revert this change
```suggestion
            tokio::time::sleep(tokio::time::Duration::from_millis(125)).await;
```

File: ank/src/cli_commands/get_state.rs [hunk 18]
```diff
@@ -440,29 +115,32 @@ mod tests {
...
-                "name2".to_string(),
-                "runtime".to_string(),
-            ),
-            generate_test_workload_spec_with_param(
-                "agent_B".to_string(),
-                "name3".to_string(),
-                "runtime".to_string(),
-            ),
-        ]);
+        // TODO: replace this and all the other calls like this one
```
Comment (krucod3) [inline ank/src/cli_commands/get_state.rs:118]: There is a todo here.

File: agent/src/runtime_manager.rs [hunk 19]
```diff
@@ -1805,22 +1802,57 @@ mod tests {
...
+                                tags: Some(ank_base::Tags {
+                                    tags: vec![ank_base::Tag {
+                                        key: "key".to_string(),
+                                        value: "value".to_string(),
+                                    }],
+                                }),
+                                runtime: Some("runtime1".to_string()),
+                                runtime_config: Some("generalOptions: [\"--version\"]\ncommandOptions: [\"--network=host\"]\nimage: alpine:latest\ncommandArgs: [\"bash\"]\n".to_string()),
+                                control_interface_access: None,
+                            },
```
Comment (krucod3) [inline agent/src/runtime_manager.rs:1836]: Wouldn't it be possible to to a generate_test... function here?

**Review by krucod3 — COMMENTED** [review]
Some small findings.

File: ank/src/cli_commands/get_state.rs (hunk 18 above)
Comment (HorjuRares) [reply ank/src/cli_commands/get_state.rs:118]: This should be treated with a specific conversion from one of the CompleteStates into a FilteredCompleteState.

Comment (krucod3) [reply ank/src/cli_commands/get_state.rs:118]: Alright, I'll handle this in my PR after this is merged in the ankaios branch

**Comment by krucod3** [comment]
There are still some warnings in the build...
Since we are not merging to Ankaios main, but to a separate branch which must again go through the build and review steps it is OK to merge like this.
I'll fix the problems in #318.
